[kn-admin]Added utils to update configmap

zhanggbj opened the pull request:
- Added utils to update configmp, as this is very common for kn-admin operations
- Chores to refine vars or logs

mattmoor commented on plugins/admin/vendor/k8s.io/apimachinery/pkg/util/wait/doc.go:
```diff
@@ -0,0 +1,19 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor commented on plugins/admin/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -0,0 +1,512 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor commented on plugins/admin/vendor/k8s.io/client-go/util/retry/util.go:
```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor replied:
<!--copyright-bot-->
Incorrect copyright year:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

mattmoor commented on plugins/admin/pkg/command/autoscaling/update.go:
```diff
@@ -17,11 +17,12 @@ package autoscaling
 import (
 	"errors"
 	"fmt"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"fmt"

```

mattmoor commented on plugins/admin/pkg/command/domain/set.go:
```diff
@@ -52,9 +58,9 @@ kn admin domain set --custom-domain mydomain.com --selector app=v1
+			currentCm, err := p.ClientSet.CoreV1().ConfigMaps(knativeServing).Get(configDomain, metav1.GetOptions{})
 			if err != nil {
-				return fmt.Errorf("failed to get configmaps: %+v", err)
+				return fmt.Errorf("failed to get ConfigMap %s in namespace %s: %+v",configDomain, knativeServing, err)
```
<!--gotool-bot-->
Format Go code:
```suggestion
				return fmt.Errorf("failed to get ConfigMap %s in namespace %s: %+v", configDomain, knativeServing, err)
```

mattmoor commented on plugins/admin/pkg/command/domain/unset.go:
```diff
@@ -17,6 +17,7 @@ package domain
 import (
 	"errors"
 	"fmt"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"fmt"

```

mattmoor commented on plugins/admin/pkg/command/domain/unset.go:
```diff
@@ -58,10 +59,12 @@ kn admin domain unset --custom-domain mydomain.com
 				return fmt.Errorf("Knative route domain %s not found\n", domain)
 			}
 
```
<!--gotool-bot-->
Format Go code:
```suggestion

```

zhanggbj commented:
@rhuss @maximilien 
The integration test failed due to incompatible packages, looks like it is not one of kn-admin, woudl you please help to take a look?
```
# k8s.io/client-go/rest
../../../pkg/mod/k8s.io/client-go@v11.0.1-0.20190805182717-6502b5e7b1b5+incompatible/rest/request.go:598:31: not enough arguments in call to watch.NewStreamWatcher
	have (*versioned.Decoder)
	want (watch.Decoder, watch.Reporter)
ERROR: cluster setup failed
```

rhuss commented:
Yep, I will give a look later today (have some meetings soon)

maximilien commented on plugins/admin/README.adoc:
```diff
@@ -120,15 +120,15 @@ Use "kn admin autoscaling [command] --help" for more information about a comman
 ----
 $ kn admin domain set --custom-domain mydomain.com
-Updated Knative route domain mydomain.com
+Set Knative route domain mydomain.com
```
Change to “... route to domain ...”

zhanggbj replied:
Same question as below.

maximilien commented on plugins/admin/README.adoc:
```diff
@@ -120,15 +120,15 @@ Use "kn admin autoscaling [command] --help" for more information about a comman
 ----
 $ kn admin domain set --custom-domain test.com --selector app=v1
-Updated Knative route domain test.com
+Set Knative route domain test.com with selector [app=v1]
```
Same as above add “to”

zhanggbj replied:
@maximilien 
As Knative supports multiple domains, I'm wondering if `set route domain to` is more like Knative only support one domain?

rhuss replied:
IMO if use "set" it implies setting the single value. Otherwise, I would expect an "Add route".

How would one manage multiple routes here ? If this is only about a 'default' route maybe "Set default Knative route domain ..." is better choice.

Also, I wonder whether the "Knative" is not redundant and maybe we should talk about a "service" instead of a "route" here, as route is an internal concept mostly hidden in kn. We are always talking about a service's or  revsion's URL.

zhanggbj replied:
@rhuss 
In fact, I think in this context it is more about `domain` other than `route`.

Knative support to have multiple domains as below, in kn-admin we want to help with add and remove domains. I'm thinking maybe `kn admin domain add` is more proper here and perhaps also change `route domain` to `domain` as it may be a little bit confusing with the Knative Route resource.

@rhuss @maximilien what do you think? Thanks!

```
data:
  # example.org will be used for routes having app=prod.
  example.org: |
    selector:
      app: prod
  # Default value for domain, for routes that does not have app=prod labels.
  # Although it will match all routes, it is the least-specific rule so it
  # will only be used if no other domain matches.
  example.com: ""
```

rhuss replied:
Yeah, the "route" confused me a bit. Maybe just referring to the domain without any revision/service prefix makes more sense. 

For this example, "Add" makes more sense as it also emphasises that you add a domain with the selector.

For general wording of the command itself (set/unset), it's a bit tricky.
As you can have multiple things that can be identified, having "create/delete" is typically a better wording for commands than "set/unset" as this more implies a single entity to manage. This is how we do it for the Knative entities like service or sources.

However, as you also can override with "set" then "add" would be misleading. Not sure if an "create/update/delete" makes sense, as this would add a possibly too fine-grained workflow (and would require the user to know whether a domain already is adder).

I hope we can implement the knative story that introduces an "apply" verb soon, which would be useful here, too: "apply" essentially mean, create-or-update and is idempotent (you can call it with the same argument many times, but only get a single result). This is the case for you command here.

So, if this were a core kn command, I probably would suggest a 

```
# Add or update the given domain with the given selector
kn admin domain apply ....

# Remove the domain
kn admin domain remove ....
```

as this aligns with the other verbs used in kn.

zhanggbj replied:
@rhuss I feel the same way. kn-admin itself should be consistent, now we have add/update/etc. This need refine and perhaps this change can wait for and follow `kn client` story you mentioned.
How about I open another chore item to follow it and we can complete this PR first.

navidshaikh replied:
IMO, set/unset suit better for properties like domain, 
while create/update/apply/deploy operates on the actual resources.

zhanggbj replied:
@navidshaikh thanks and we can continue subcommand wording in #53

maximilien commented on plugins/admin/pkg/command/autoscaling/update.go:
```diff
@@ -38,8 +40,8 @@ var (
-		Short: "update autoscaling config",
-		Long: `update autoscaling config provided by Knative Pod Autoscaler (KPA)
+		Short: "Update autoscaling config",
+		Long: `Update autoscaling config provided by Knative Pod Autoscaler (KPA)
```
I kind of expected the Long description to have an example...

zhanggbj replied:
Yes, we do have examples, there's a new line here

```
For example:
# To enable scale-to-zero
kn admin autoscaling update --scale-to-zero
# To disable scale-to-zero
kn admin autoscaling update --no-scale-to-zero
```

maximilien commented on plugins/admin/pkg/command/domain/set.go:
```diff
@@ -19,22 +19,28 @@ import (
-var selector []string
-var domain string
+var (
+	ScaleToZero    bool
```
Why is `ScaleToZero` var public when the others are private?

zhanggbj replied:
Thanks Max! this is typo and fixed.

maximilien commented on plugins/admin/pkg/command/registry/add.go:
```diff
@@ -124,9 +124,9 @@ kn admin registry add \
 			_, err = p.ClientSet.CoreV1().ServiceAccounts("default").Update(desiredSa)
 			if err != nil {
-				return fmt.Errorf("failed to add registry secret in default serviceaccount: %v", err)
+				return fmt.Errorf("Failed to add registry secret in default serviceaccount: %v", err)
```
Why is the message capitalized here but lower case before. Let’s be consistent.

zhanggbj replied:
Thanks Max! Fixed.

mattmoor commented on plugins/admin/pkg/command/utils/utils_test.go:
```diff
@@ -0,0 +1,93 @@
+
+package utils
+
+import (
```
<!--gotool-bot-->
Format Go code:
```suggestion
import (
	"testing"

```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
import (
	"testing"

```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
import (
	"testing"

```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
import (
	"testing"

```

mattmoor commented on plugins/admin/pkg/command/utils/utils_test.go:
```diff
@@ -0,0 +1,93 @@
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	"testing"
```
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion

```

zhanggbj commented:
@maximilien 
I addressed your comments, and would you please help to take a look? Thanks!
- e2e test case not ready yet, covered by #14 
- Kpow integration tests failures due to repo settings (Roland is taking care)

zhanggbj commented:
@rhuss 
Thanks! As the example here, kn-admin will also add the domains with/without the selector. When users need to remove a domain, they only need to provide the domain name.
And `kn admin domain list` is working in progress, to list all the added domains and with selectors if any.

mattmoor commented on plugins/admin/pkg/command/autoscaling/autoscaling.go:
```diff
@@ -24,9 +24,7 @@ func NewAutoscalingCmd(p *pkg.AdminParams) *cobra.Command {
-		Long: `Manage autoscaling provided by Knative Pod Autoscaler (KPA). For example:
-
-kn admin autoscaling update - to manage autoscaling config`,
+		Long: `Manage autoscaling provided by Knative Pod Autoscaler (KPA).`,
```
<!--gotool-bot-->
Format Go code:
```suggestion
		Long:  `Manage autoscaling provided by Knative Pod Autoscaler (KPA).`,
```

mattmoor commented on plugins/admin/pkg/command/autoscaling/update.go:
```diff
@@ -38,49 +40,46 @@ var (
-kn admin autoscaling update --no-scale-to-zero
-`,
+		Short: "Update autoscaling config",
+		Long: `Update autoscaling config provided by Knative Pod Autoscaler (KPA)`,
```
<!--gotool-bot-->
Format Go code:
```suggestion
		Long:  `Update autoscaling config provided by Knative Pod Autoscaler (KPA)`,
```

mattmoor commented on plugins/admin/pkg/command/domain/domain.go:
```diff
@@ -24,11 +24,7 @@ func NewDomainCmd(p *pkg.AdminParams) *cobra.Command {
-kn admin domain set - to set Knative route domain
-
-kn admin domain unset - to unset Knative route domain`,
+		Long: `Manage default route domain or custom route domain for Service with selectors.`,
```
<!--gotool-bot-->
Format Go code:
```suggestion
		Long:  `Manage default route domain or custom route domain for Service with selectors.`,
```

mattmoor commented on plugins/admin/pkg/command/domain/set.go:
```diff
@@ -19,30 +19,34 @@ import (
-kn admin domain set --custom-domain mydomain.com --selector app=v1
-`,
+		Short: "Set route domain",
+		Long: `Set Knative route domain for service`,
+		Example:`
```
<!--gotool-bot-->
Format Go code:
```suggestion
		Long:  `Set Knative route domain for service`,
		Example: `
```

mattmoor commented on plugins/admin/pkg/command/domain/unset.go:
```diff
@@ -29,13 +31,11 @@ import (
-kn admin domain unset --custom-domain mydomain.com
-`,
+		Short: "Unset route domain",
+		Long: `Unset Knative route domain for service`,
```
<!--gotool-bot-->
Format Go code:
```suggestion
		Long:  `Unset Knative route domain for service`,
```

mattmoor commented on plugins/admin/pkg/command/registry/add.go:
```diff
@@ -56,16 +56,16 @@ var prflags prcmdFlags
-  --username=[REGISTRY_USER] \
-  --password=[REGISTRY_PASSWORD]`,
+		Short: "Add registry with credentials",
+		Long: `Add registry with credentials to enable Service deployment from it`,
+ 		Example: `
```
<!--gotool-bot-->
Format Go code:
```suggestion
		Long:  `Add registry with credentials to enable Service deployment from it`,
		Example: `
```

mattmoor commented on plugins/admin/pkg/command/registry/registry.go:
```diff
@@ -24,15 +24,7 @@ func NewPrivateRegistryCmd(p *pkg.AdminParams) *cobra.Command {
-  --email=[REGISTRY_EMAIL] \
-  --username=[REGISTRY_USER] \
-  --password=[REGISTRY_PASSWORD]`,
+		Long: `Manage Service deployment from a registry with credentials`,
```
<!--gotool-bot-->
Format Go code:
```suggestion
		Long:  `Manage Service deployment from a registry with credentials`,
```

zhanggbj commented:
@rhuss As discussed in https://github.com/knative/client-contrib/pull/40#discussion_r442624507,  I opened another chore item to follow command wording create/apply/update (#53). Please help to review again and hope we can merge this first before the vendor pkg change. thanks!

rhuss commented:
/hold

rhuss commented:
@maximilien please unhold if this looks ok for you, so that it gets merged.

zhanggbj commented:
@maximilien Thanks Max. For test, we have UT for each command and also for [registry/add](https://github.com/knative/client-contrib/blob/master/plugins/admin/pkg/command/registry/add_test.go)
e2e test is still working in progress by #14

maximilien commented:
> @maximilien Thanks Max. For test, we have UT for each command and also for [registry/add](https://github.com/knative/client-contrib/blob/master/plugins/admin/pkg/command/registry/add_test.go)
> e2e test is still working in progress by #14

OK cool. Let me know when you address the last comments from @navidshaikh and we can merge this. Thx.

rhuss commented:
/unhold

zhanggbj commented on plugins/admin/pkg/command/autoscaling/update.go:
```diff
@@ -39,49 +40,46 @@ var (
+		Long:  `Update autoscaling config provided by Knative Pod Autoscaler (KPA)`,
+		Example: `
+  # To enable scale-to-zero
+  kn admin autoscaling update --scale-to-zero
+  # To disable scale-to-zero
```
@navidshaikh yes, I agree, this is a known issue and will be  handled https://github.com/knative/client-contrib/issues/41

zhanggbj commented on plugins/admin/pkg/command/domain/set.go:
```diff
@@ -19,30 +19,34 @@ import (
+  # To set a default route domain
+  kn admin domain set --custom-domain mydomain.com
+
+  # To set a route domain for service having label app=v1
```
@navidshaikh Thanks for the suggestion. Will update in another PR to refine https://github.com/knative/client-contrib/issues/41
